draft-ietf-quic-tls.md

```diff
@@ -1007,144 +813,94 @@ encrypted_pn = ChaCha20(pn_key, counter, nonce, packet_number)
-endpoint use packets protected with handshake keys.
-
-Any TLS handshake messages that are sent after completing the TLS handshake do
-not need special packet protection rules.  Packets containing these messages use
-the packet protection keys that are current at the time of sending (or
-retransmission).
-
-Like the client, a server MUST send retransmissions of its unprotected handshake
-messages or acknowledgments for unprotected handshake messages sent by the
-client in packets protected with handshake keys.
-
-
-### Initial Key Transitions {#first-keys}
-
-Once the TLS handshake is complete, keying material is exported from TLS and
-used to protect QUIC packets.
+## Use of 0-RTT Keys {#using-early-data}
 
-Packets protected with 1-RTT keys initially have a KEY_PHASE bit set to 0.  This
-bit inverts with each subsequent key update (see {{key-update}}).
+If 0-RTT keys are available (see {{enable-0rtt}}), the lack of replay protection
+means that restrictions on their use are necessary to avoid replay attacks on
+the protocol.
 
-If the client sends 0-RTT data, it uses the 0-RTT packet type.  The packet that
-contains the TLS EndOfEarlyData and Finished messages are sent in packets
-protected with handshake keys.
+A client MUST only use 0-RTT keys to protect data that is idempotent.  A client
+MAY wish to apply additional restrictions on what data it sends prior to the
+completion of the TLS handshake.  A client otherwise treats 0-RTT keys as
+equivalent to 1-RTT keys, except that it ACKs MUST only be sent with
+1-RTT keys.
 
-Using distinct packet types during the handshake for handshake messages, 0-RTT
-data, and 1-RTT data ensures that the server is able to distinguish between the
-different keys used to remove packet protection.  All of these packets can
-arrive concurrently at a server.
+A client that receives an indication that its 0-RTT data has been accepted by a
+server can send 0-RTT data until it receives all of the server's handshake
+messages.  A client SHOULD stop sending 0-RTT data if it receives an indication
+that 0-RTT data has been rejected.
 
-A server might choose to retain 0-RTT packets that arrive before a TLS
-ClientHello.  The server can then use those packets once the ClientHello
-arrives.  However, the potential for denial of service from buffering 0-RTT
-packets is significant.  These packets cannot be authenticated and so might be
-employed by an attacker to exhaust server resources.  Limiting the number of
-packets that are saved might be necessary.
+A server MUST NOT use 0-RTT keys to protect packets.
 
-The server transitions to using 1-RTT keys after sending its first flight of TLS
-handshake messages, ending in the Finished.
-From this point, the server protects all packets with 1-RTT
-keys.  Future packets are therefore protected with 1-RTT keys.  Initially, these
-are marked with a KEY_PHASE of 0.
+If a server rejects 0-RTT, then the TLS stream will not include any TLS records
+protected with 0-RTT keys.
```

Comment:
The TLS stream will not include any records protected with 0-RTT keys anyway, because the records on the TLS stream aren't protected anymore.  This statement seems somewhere between redundant and misleading.

Reply:
It's actually Handshake messages, but I agree it's confusing, so I'm removing it.